db/pushdb.go

```diff
@@ -145,11 +146,42 @@ func (f *pushDatabaseOpts) AddPushServiceProviderToService(service string,
 	}
 	f.dblock.Lock()
 	defer f.dblock.Unlock()
+
+	//#198:
+	//2017 Victor Lang
+	//Before add a new psp to service, try to verify there is no redundant psp in same type exists under the same service.
+	//Currently, redundant psp to service will result in problem in subscribe and push later.
+	expsps, err := f.db.GetPushServiceProvidersByService(service)
+	if err != nil {
+		return fmt.Errorf("Error querying psp with name: %v", err)
+	}
+	if len(expsps) > 0 {
+		for index := range expsps {
+			pushpsp, perr := f.db.GetPushServiceProvider(expsps[index])
+
+			if perr != nil {
+				return fmt.Errorf("Error to retrieve psp with name: %v", err)
+			}
+			//Whether the existing psp has the same push service type
+			if pushpsp.PushServiceName() == push_service_provider.PushServiceName() {
+				//The service has a psp in same push service type
+				//The same fixed data are allowed under this situation in case the user wants to update "VolatileData" data in psp,
+				// and the "different" fixed data is disallowed.
+				//Since the psp's fixed data currently is used to generate UNIQUE pushpeer name, let's directly compare the Name() of pushpeer for now.
+				if pushpsp.PushPeer.Name() != push_service_provider.PushPeer.Name() {
+					return fmt.Errorf("Provider with different fixed data already exists in the service %s for type %s. The detail of new psp: %v", service, push_service_provider.PushServiceName(), push_service_provider)
```

Comment:
Mention remediation steps in the error message

the fixed data (cert, key file path for APNS) are the same,
 the solution would be to change contents of any cert file
Alternately, this could be fixed by removing the old PSP, but that would delete subscribers